pkg/gadgets/mountsnoop/gadget.go

```diff
@@ -0,0 +1,147 @@
+	"encoding/json"
+	"fmt"
+
+	log "github.com/sirupsen/logrus"
+
+	"github.com/kinvolk/inspektor-gadget/pkg/gadgets"
+	"github.com/kinvolk/inspektor-gadget/pkg/gadgets/mountsnoop/tracer"
+	coretracer "github.com/kinvolk/inspektor-gadget/pkg/gadgets/mountsnoop/tracer/core"
+	standardtracer "github.com/kinvolk/inspektor-gadget/pkg/gadgets/mountsnoop/tracer/standard"
+
+	"github.com/kinvolk/inspektor-gadget/pkg/gadgets/mountsnoop/types"
+
+	gadgetv1alpha1 "github.com/kinvolk/inspektor-gadget/pkg/apis/gadget/v1alpha1"
+)
+
+type Trace struct {
+	resolver gadgets.Resolver
+
+	started bool
+	tracer  tracer.Tracer
+}
+
+type TraceFactory struct {
+	gadgets.BaseFactory
+}
+
+func NewFactory() gadgets.TraceFactory {
+	return &TraceFactory{
+		BaseFactory: gadgets.BaseFactory{DeleteTrace: deleteTrace},
+	}
+}
+
+func (f *TraceFactory) Description() string {
+	return `mountsnoop traces mount and umount syscalls`
+}
+
+func (f *TraceFactory) OutputModesSupported() map[string]struct{} {
+	return map[string]struct{}{
+		"Stream": {},
+	}
+}
+
+func deleteTrace(name string, t interface{}) {
+	trace := t.(*Trace)
+	if trace.tracer != nil {
+		trace.tracer.Stop()
+	}
+}
+
+func (f *TraceFactory) Operations() map[string]gadgets.TraceOperation {
+	n := func() interface{} {
+		return &Trace{
+			resolver: f.Resolver,
+		}
+	}
+
+	return map[string]gadgets.TraceOperation{
+		"start": {
+			Doc: "Start mountsnoop gadget",
+			Operation: func(name string, trace *gadgetv1alpha1.Trace) {
+				f.LookupOrCreate(name, n).(*Trace).Start(trace)
+			},
+		},
+		"stop": {
+			Doc: "Stop mountsnoop gadget",
+			Operation: func(name string, trace *gadgetv1alpha1.Trace) {
+				f.LookupOrCreate(name, n).(*Trace).Stop(trace)
+			},
+		},
+	}
+}
+
+func (t *Trace) Start(trace *gadgetv1alpha1.Trace) {
+	if t.started {
+		gadgets.CleanupTraceStatus(trace)
+		trace.Status.State = "Started"
+		return
+	}
+
+	traceName := gadgets.TraceName(trace.ObjectMeta.Namespace, trace.ObjectMeta.Name)
+
+	eventCallback := func(event types.Event) {
+		r, err := json.Marshal(event)
+		if err != nil {
+			log.Warnf("Gadget %s: error marshalling event: %s", trace.Spec.Gadget, err)
+			return
+		}
+		t.resolver.PublishEvent(traceName, string(r))
+	}
+
+	var err error
+
+	config := &tracer.Config{
+		MountnsMap: gadgets.TracePinPath(trace.ObjectMeta.Namespace, trace.ObjectMeta.Name),
+	}
+	t.tracer, err = coretracer.NewTracer(config, t.resolver, eventCallback, trace.Spec.Node)
+	if err != nil {
+		// TODO: The following line causes the client to fail.
+		trace.Status.OperationWarning = fmt.Sprint("failed to create core tracer. Falling back to standard one")
+
+		// fallback to standard tracer
+		log.Infof("Gadget %s: falling back to standard tracer. CO-RE tracer failed: %s",
+			trace.Spec.Gadget, err)
+
+		t.tracer, err = standardtracer.NewTracer(config, t.resolver, eventCallback, trace.Spec.Node)
+		if err != nil {
+			trace.Status.OperationError = fmt.Sprintf("failed to create tracer: %s", err)
```

Comment:
```suggestion
			gadgets.CleanupTraceStatus(trace)
			trace.Status.OperationError = fmt.Sprintf("failed to create tracer: %s", err)
```

Reply:
This is not consistent among different gadgets. I opened https://github.com/kinvolk/inspektor-gadget/issues/522 to take care of it.

Reply:
Just remember to update this once https://github.com/kinvolk/inspektor-gadget/pull/528 is merged.